include/navigation/image_database.h

```diff
@@ -73,78 +79,245 @@ class ImageDatabase
+            m_Recording = false;
+        }
 
         //! Current number of *new* entries for the ImageDatabase
-        size_t size() const;
+        size_t size() const { return m_NewEntries.size(); }
 
         //! Get the format in which images will be saved
-        std::string getImageFormat() const;
+        std::string getImageFormat() const { return m_ImageFormat; }
+
+        void setImageFormat(const std::string &imageFormat)
+        {
+            m_ImageFormat = imageFormat;
+        }
 
     private:
+        const std::array<std::string, NumExtraFields> m_ExtraFieldNames;
+        std::string m_ImageFormat = "png";
         ImageDatabase &m_ImageDatabase;
-        const std::string m_ImageFormat;
         bool m_Recording;
         std::vector<Entry> m_NewEntries;
 
     protected:
         cv::FileStorage m_YAML;
 
+        template<class... Ts>
         Recorder(ImageDatabase &imageDatabase,
                  const bool isRoute,
-                 const std::string &imageFormat);
+                 Ts &&... extraFieldNames)
+          : m_ExtraFieldNames({ std::forward<Ts>(extraFieldNames)... })
+          , m_ImageDatabase(imageDatabase)
+          , m_Recording(true)
+          , m_YAML(".yml", cv::FileStorage::WRITE | cv::FileStorage::MEMORY)
+        {
+            // Set this property of the ImageDatabase
+            imageDatabase.m_IsRoute = isRoute;
+
+            // Get current date and time
+            std::time_t now = std::time(nullptr);
+            char timeStr[sizeof("0000-00-00 00:00:00")];
+            BOB_ASSERT(0 != std::strftime(timeStr, sizeof(timeStr), "%F %T", std::localtime(&now)));
+
+            // Write some metadata; users can add extra
+            m_YAML << "metadata"
+                   << "{"
+                   << "time" << timeStr
+                   << "project_git_commit" << BOB_PROJECT_GIT_COMMIT
+                   << "bob_robotics_git_commit" << BOB_ROBOTICS_GIT_COMMIT
+                   << "type" << (isRoute ? "route" : "grid");
+        }
 
+        template<class... Ts>
         void addEntry(const std::string &filename,
                       const cv::Mat &image,
                       const Vector3<millimeter_t> &position,
                       const degree_t heading,
-                      const std::array<size_t, 3> &gridPosition = { 0, 0, 0 });
+                      const std::array<size_t, 3> &gridPosition,
+                      Ts&&... extraFieldValues)
+        {
+            static_assert(sizeof...(extraFieldValues) == NumExtraFields,
+                          "Must supply correct number of extra field values");
+            BOB_ASSERT(m_Recording);
+
+            m_ImageDatabase.writeImage(filename, image);
+
+            Entry newEntry{
+                position,
+                heading,
+                m_ImageDatabase.m_Path / filename,
+                gridPosition,
+                {}
+            };
+            m_NewEntries.emplace_back(std::move(newEntry));
```

Comment:
Does this not work here:
```c++
m_NewEntries.emplace_back(
    position,
    heading,
    m_ImageDatabase.m_Path / filename,
    gridPosition,
    {});
```
?
Otherwise I think you should ``push_back`` here. ``newEntry`` is on the stack so it can't be 'moved' in any useful way to a heap data structure. I'm a little surprised you don't get a warning here (that may only be a clang-level OCD thing)

Reply:
> Does this not work here:
> 
> ```c++
> m_NewEntries.emplace_back(
>     position,
>     heading,
>     m_ImageDatabase.m_Path / filename,
>     gridPosition,
>     {});
> ```
> 
> ?
For some reason this didn't work for me, though I feel it should work too.

> Otherwise I think you should `push_back` here. `newEntry` is on the stack so it can't be 'moved' in any useful way to a heap data structure. I'm a little surprised you don't get a warning here (that may only be a clang-level OCD thing)
I *think* ``Entry`` will have a default move ctor, which just moves its moveable members (i.e. the filepath and the ``std::unordered_map``) and copies the others. So it might avoid a copy for the filepath this way.

Reply:
Ahh maybe you need to use ``std::piecewise_construct`` and ``std::forward_as_tuple`` to deal with the {}s

Reply:
That sounds sensible... but I can't seem to get it to work here :thinking:. I've tried various permutations of brackets/not-brackets etc. but the compiler keeps complaining. Any thoughts?

Reply:
Ahh not worth it then!